### PR TITLE
Spark DataSource V2 Write with "BigQuery Storage Write API"

### PR DESCRIPTION
**Pull request by himanshukohli09**
Co-authored-by: Yuval Medina <yuvalmedina@users.noreply.github.com>

### CODE REVIEW

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java [hunk 1]
```diff
@@ -163,21 +165,21 @@ public void testWriteToBigQuery_AppendSaveMode() {
...
-  public void testWriteToBigQuery_ErrorIfExistsSaveMode() {
+  public void testWriteToBigQuery_ErrorIfExistsSaveMode() throws InterruptedException {
     // initial write
     writeToBigQuery(initialData(), SaveMode.ErrorIfExists);
     assertThat(testTableNumberOfRows()).isEqualTo(2);
     assertThat(initialDataValuesExist()).isTrue();
     // second write
-    assertThrows(
-        IllegalArgumentException.class,
+    Assert.assertThrows(
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java]: Done

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java [hunk 2]
```diff
@@ -163,21 +165,21 @@ public void testWriteToBigQuery_AppendSaveMode() {
...
+        Exception.class,
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java]: Its tricky! V1 and V2 throws different exceptions. V1 throws IllegalArgumentException and V2 throws ProvisionException as the code breaks at guice.

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java [hunk 3]
```diff
@@ -379,21 +400,37 @@ private StructType structType(StructField... fields) {
...
+    if (isDirectWrite) {
+      return;
+    }
     testPartition("HOUR");
   }
 
   @Test
   public void testPartition_Daily() {
+    // partition write not supported in BQ Storage Write API
+    if (isDirectWrite) {
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java]: Done

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java [hunk 4]
```diff
@@ -379,21 +400,37 @@ private StructType structType(StructField... fields) {
...
+      return;
+    }
     testPartition("DAY");
   }
 
   @Test
   public void testPartition_Monthly() {
+    // partition write not supported in BQ Storage Write API
+    if (isDirectWrite) {
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java]: Done

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java [hunk 5]
```diff
@@ -379,21 +400,37 @@ private StructType structType(StructField... fields) {
...
+      return;
+    }
     testPartition("MONTH");
   }
 
   @Test
   public void testPartition_Yearly() {
+    // partition write not supported in BQ Storage Write API
+    if (isDirectWrite) {
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java]: Done

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java [hunk 6]
```diff
@@ -379,21 +400,37 @@ private StructType structType(StructField... fields) {
 
   @Test
   public void testPartition_Hourly() {
+    // partition write not supported in BQ Storage Write API
+    if (isDirectWrite) {
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java]: Done

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java [hunk 7]
```diff
@@ -15,8 +15,367 @@
...
+    }
+  }
+
+  static String sqlFromFormat(
+      String queryFormat, TableId destinationTableId, TableId temporaryTableId) {
+    return String.format(
+        queryFormat, fullTableName(destinationTableId), fullTableName(temporaryTableId));
+  }
+
+  static String fullTableName(TableId tableId) {
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java:159]: The method in super is different than this.

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java [hunk 8]
```diff
@@ -15,8 +15,367 @@
...
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.threeten.bp.Duration;
+
 public class Spark24WriteIntegrationTest extends WriteIntegrationTestBase {
 
-  // tests are from the super-class
+  public Spark24WriteIntegrationTest() {
+    super(true);
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java]: Done

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java:168]: It exists in BigQueryClient. make it public static.

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/integration/Spark24WriteIntegrationTest.java]: Done!

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**